### ISSUE TITLE
[Truffle-v5] Wrong return type for functions with multiple named return values

### ISSUE THREAD

**Issue by fleupold** [ethereum-ts/TypeChain#232]
Adding the following function to the `MetaCoin` contract in `examples/truffle-v5`:

```solidity
function foo() public pure returns (uint bar, address bam) {
  bar = 42;
  bam = address(0);
}
```

generates the following type bindings:

```ts
foo(txDetails?: Truffle.TransactionDetails): Promise<[BN, string]>;
```

Therefore, accessing these values in a test like this compiles fine:

```ts
it('should allow calling foo', async () => {
  const metaCoinInstance = await MetaCoin.new();
  let [bar, bam] = await metaCoinInstance.foo();
  assert.equal(bar.toNumber(), 42);
});
```

However, running the test crashes with the following Type error:

> 1) Contract: MetaCoin
  should allow calling foo:
TypeError: (intermediate value) is not iterable
at Context.<anonymous> (test/metacoin.ts:43:22)
at processTicksAndRejections (internal/process/task_queues.js:97:5)

I believe the reason for this is that truffle v5 uses web3 v1 under the hood and thus the function above should actually compile to a Promise of a named struct rather than a Promise of a tuple. 
In this case the method should have the following signature:

```ts
foo(txDetails?: Truffle.TransactionDetails): Promise<{
  bar: BN;
  bam: string;
  0: BN;
  1: string;
}>;
```

(which is already the case when using `--target=web3-v1`)

As a temporary work-around, accessing the field by index compiles fine and does not crash (since the named struct also has a field with name 0 & 1):

```ts
  let result = await metaCoinInstance.foo();
  assert.equal(result[0], 42);
```

**Comment by krzkaczor** [comment]
Thanks! This seems like a bug. I am surprised it wasn't caught by our test suite 🤔

**Comment by JustynaBroniszewska** [comment]
I might have a similar issue. I created a repo that shows it [here](https://github.com/JustynaBroniszewska/typechain-issue/blob/master/index.ts#L10)

TL;DR 
When I call a function, that is supposed to return multiple values I have a problem destructuring them: 
```java
contract Example {
    uint a = 1;
    uint b = 2;

    function getNumbers() public view returns (uint, uint) {
        return (a, b);
    }
}
```
This doesn't work:
```ts
  const [a, b] = await contract.getNumbers()
```
...but these work:
```ts
  const { 0: a, 1: b } = await contract.getNumbers()
  const numbers = await contract.getNumbers()
  console.log(numbers[0], numbers[1])
```

**Comment by quezak** [comment]
@JustynaBroniszewska you're right! Looks to be a simple fix, would you like to prepare a PR? (plus a test so we catch it if it changes again :wink: )
([relevant code location](https://github.com/ethereum-ts/TypeChain/blob/46ed6338d94add22cb1bdb94f6b4a1f01175878a/packages/target-truffle-v5/src/codegen/types.ts#L16))

**Comment by krzkaczor** [comment]
I managed to create a failing test case: https://github.com/ethereum-ts/TypeChain/pull/311

Tmrw, I'll work on a fix. Probably it's a bug in all targets.

**Comment by krzkaczor** [comment]
Released: https://github.com/ethereum-ts/TypeChain/releases/tag/typechain%404.0.1
